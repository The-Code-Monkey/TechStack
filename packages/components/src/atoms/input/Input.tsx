import { useCallback, useState, useContext } from 'react';
import { ThemeContext } from 'styled-components';

import { BoxProps } from '../../primal';
import { DefaultThemeWithDefaultStyles } from '../../utils';

import Checkbox from './checkbox';
import { StyledInput } from './styled';
import { InputPropsUnion } from './types';

export type Props = BoxProps & InputPropsUnion;

const Input = ({ onChange, type, value, name, ...rest }: Props) => {
  const theme = useContext<DefaultThemeWithDefaultStyles>(ThemeContext);
  const [v, setValue] = useState<string | number | undefined | unknown>(value);

  const handleOnChange = useCallback(
    (event: any) => {
      if (typeof event !== 'boolean') {
        setValue(event.target.value);
      } else {
        setValue(event);
      }
      if (onChange) onChange(event);
    },
    [onChange]
  );

  const renderInput = useCallback(
    (type: InputPropsUnion['type']) => {
      switch (type) {
        case 'checkbox': {
          return (
            <Checkbox
              name={name}
              value={value}
              onChange={handleOnChange}
              {...theme.defaultStyles?.checkbox}
              {...rest}
            />
          );
        }
        default: {
          return (
            <StyledInput
              name={name}
              value={v}
              type={type}
              onChange={handleOnChange}
              {...theme.defaultStyles?.input}
              {...rest}
            />
          );
        }
      }
    },
    [type]
  );

  return renderInput(type);
};

export default Input;
