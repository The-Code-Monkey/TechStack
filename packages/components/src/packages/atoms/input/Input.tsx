import { css } from '@aw-web-design/styled-system';
import React, {
  useCallback,
  ChangeEvent,
  HTMLInputTypeAttribute,
  useState,
  useContext,
} from 'react';
import { ThemeContext } from 'styled-components';

import { BoxProps } from '../../primatives';
import { ITheme } from '../../utils';

import { StyledInput } from './styled';

export interface Props extends BoxProps {
  onChange?: (e: ChangeEvent<HTMLInputElement>) => void;
  type?: HTMLInputTypeAttribute;
  value?: string | number;
  name: string;
  placeholder?: string;
}

const Input = ({
  onChange,
  type = 'text',
  value,
  name,
  placeholder = 'Placeholder',
  ...rest
}: Props) => {
  const theme = useContext<ITheme<unknown>>(ThemeContext);
  const [v, setValue] = useState<string | number | undefined>(value);

  const handleOnChange = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      setValue(event.target.value);
      if (onChange) onChange(event);
    },
    [onChange]
  );

  console.log(theme);

  console.log(
    css({
      ...(theme.defaultStyles.input as object),
    })()
  );

  return (
    <StyledInput
      name={name}
      placeholder={placeholder}
      value={v}
      type={type}
      onChange={handleOnChange}
      {...theme.defaultStyles.input}
      {...rest}
    />
  );
};

export default Input;
