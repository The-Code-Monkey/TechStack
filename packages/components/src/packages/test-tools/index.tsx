import { mount, shallow, ShallowWrapper, ReactWrapper } from 'enzyme';
import React, { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components';

import theme from '../theme';
import { ConfigContext } from '../utils';

const config = {
  outputDir: '/src/packages/',
  iconDir: 'feather',
};

const removeProperties = () => {
  const keys = ['theme', 'styledTheme'];
  return {
    test: (val: any) =>
      val &&
      typeof val === 'object' &&
      'props' in val &&
      Object.keys(val.props).some((prop) => keys.some((key) => key === prop)),
    print: (val: any, serialize: any) => {
      keys.forEach((key) => {
        delete val.props[key];
      });
      return serialize(val);
    },
  };
};

const formattedTheme = {
  ...theme,
  colors: {
    common: theme.colors?.common,
    modes: theme.colors?.modes,
    ...theme.colors?.modes.light,
  },
};

export const shallowWithTheme: (children: ReactNode) => ShallowWrapper = (
  children: ReactNode
) =>
  shallow(<ThemeProvider theme={formattedTheme}>{children}</ThemeProvider>)
    .dive()
    .shallow();

export const mountWithTheme: (children: ReactNode) => ReactWrapper = (
  children: ReactNode
) => {
  expect.addSnapshotSerializer(removeProperties());

  return mount(
    <ConfigContext.Provider value={config}>
      <ThemeProvider theme={formattedTheme}>
        <>{children}</>
      </ThemeProvider>
    </ConfigContext.Provider>
  );
};
