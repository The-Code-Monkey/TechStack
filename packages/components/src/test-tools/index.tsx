import isPropValid from '@emotion/is-prop-valid';
import { render, RenderResult } from '@testing-library/react';
import { ReactNode } from 'react';
import { StyleSheetManager, ThemeProvider } from 'styled-components';
import { expect } from 'bun:test';

import theme from '../theme';
import { ConfigContext } from '../utils';

const config = {
  outputDir: '/src/packages/',
  iconDir: 'feather',
};

const removeProperties = () => {
  const keys = ['theme', 'styledTheme'];
  return {
    test: (val: { props: unknown }) =>
      val &&
      typeof val === 'object' &&
      'props' in val &&
      Object.keys(val.props).some(prop => keys.some(key => key === prop)),
    serialize: (val, config, indentation, depth, refs, printer) => {
      keys.forEach(key => {
        delete val.props[key];
      });
      return printer(val, config, indentation, depth, refs);
    },
  };
};

const formattedTheme = {
  ...theme,
  colors: {
    common: theme.colors?.common,
    modes: theme.colors?.modes,
    ...theme.colors?.modes.light,
  },
};

export const mountWithTheme = (children: ReactNode): RenderResult => {
  // expect.addSnapshotSerializer(removeProperties());

  return render(
    <StyleSheetManager shouldForwardProp={isPropValid}>
      <ConfigContext.Provider value={config}>
        <ThemeProvider theme={formattedTheme}>
          <>{children}</>
        </ThemeProvider>
      </ConfigContext.Provider>
    </StyleSheetManager>
  );
};
