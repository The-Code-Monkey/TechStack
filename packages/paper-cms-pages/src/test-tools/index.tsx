import { ConfigContext } from '@techstack/components';
import { render } from '@testing-library/react';
import { ReactNode } from 'react';

import theme from '../theme';
import { ThemeProvider } from '../workarounds/styled-components';

const config = {
  outputDir: '/src/packages/',
  iconDir: 'feather',
};

const removeProperties = () => {
  const keys = ['theme', 'styledTheme'];
  return {
    test: (val: { props: unknown }) =>
      val &&
      typeof val === 'object' &&
      'props' in val &&
      Object.keys(val.props).some(prop => keys.some(key => key === prop)),
    serialize: (val, config, indentation, depth, refs, printer) => {
      keys.forEach(key => {
        delete val.props[key];
      });
      return printer(val, config, indentation, depth, refs);
    },
  };
};

const formattedTheme = {
  ...theme,
  colors: {
    common: theme.colors?.common,
    modes: theme.colors?.modes,
    ...theme.colors?.modes.light,
  },
};

export const mountWithTheme = (children: ReactNode) => {
  expect.addSnapshotSerializer(removeProperties());

  return render(
    <ConfigContext.Provider value={config}>
      <ThemeProvider theme={formattedTheme}>
        <>{children}</>
      </ThemeProvider>
    </ConfigContext.Provider>
  );
};
