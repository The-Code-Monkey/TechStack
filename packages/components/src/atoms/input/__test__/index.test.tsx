import { mountWithTheme } from '../../../test-tools';
import Input from '../index';

describe('<Input />', () => {
  it('renders correctly', () => {
    const { asFragment } = mountWithTheme(
      <Input name='test' placeholder='Placeholder' />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('textarea renders correctly', () => {
    const { asFragment } = mountWithTheme(
      <Input name='test' placeholder='Placeholder' type='textarea' />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('email renders correctly', () => {
    const { asFragment } = mountWithTheme(
      <Input name='test' placeholder='Placeholder' type='email' />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly', () => {
    const { asFragment } = mountWithTheme(
      <Input name='test' placeholder='Placeholder' type='password' />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('checkbox renders correctly', () => {
    const { asFragment } = mountWithTheme(
      <Input name='test' type='checkbox' onChange={(a) => console.log(typeof a === "boolean")} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
