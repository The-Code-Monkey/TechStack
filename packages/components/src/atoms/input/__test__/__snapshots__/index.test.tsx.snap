// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> checkbox renders correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-kvAqsT ghVUiJ sc-cdtKKh bQZlMA"
    data-checked="false"
    data-indeterminate="false"
    data-testid="checkbox"
    name="test"
  />
</DocumentFragment>
`;

exports[`<Input /> email renders correctly 1`] = `
<DocumentFragment>
  <input
    class="sc-cLrhaH cbZXBu"
    cursor="pointer"
    name="test"
    placeholder="Placeholder"
    type="email"
    width="100%"
  />
</DocumentFragment>
`;

exports[`<Input /> renders correctly 1`] = `
<DocumentFragment>
  <input
    class="sc-cLrhaH cbZXBu"
    cursor="pointer"
    name="test"
    placeholder="Placeholder"
    width="100%"
  />
</DocumentFragment>
`;

exports[`<Input /> renders correctly 2`] = `
<DocumentFragment>
  <input
    class="sc-cLrhaH cbZXBu"
    cursor="pointer"
    name="test"
    placeholder="Placeholder"
    type="password"
    width="100%"
  />
</DocumentFragment>
`;

exports[`<Input /> textarea renders correctly 1`] = `
<DocumentFragment>
  <textarea
    class="sc-jWWaec jWWOkQ"
    cursor="pointer"
    name="test"
    placeholder="Placeholder"
    width="100%"
  />
</DocumentFragment>
`;
